### ISSUE TITLE
Add project bins/Asset manager

### ISSUE THREAD

**Issue by Nasosecco** [audacity/audacity#2462]
Hi everyone,

First of all thank you for your work: I use Audacity since many years with great satisfaction. 
I am a radio journalist and I would like to work with Audacity, which is a very powerful tool and whose Open Source philosophy I share. But unfortunately as soon as I have to work on complex projects involving the editing of a broadcast composed of numerous interviews the use of audacity becomes inconvenient.

I try to explain myself in more detail by giving an example of my workflow.

Normally a 25-minute audio documentary consists of 2-3 long interviews (10 minutes to an hour), music, archival pieces and sound effects.

My editing workflow consists of several steps:

1. Transcription of the interviews
2. Selection in each interview of the parts that will be used in the final edit
3. Dividing the interview into segments that will then be organized along with music and sound effects in the final edit
4. Final editing in a single file
5. Mixing, mastering, etc.

The difficulty in using Audacity lies in the fact that there is no effective way to perform the actions in point 2 and 3.
In the programs I normally use (HIndenburg Journalist, DigaSystem) it is possible to select directly in the waveform a part of an interview and add it to a clipboard where all the clips of an interview can be organized to be reused later and rearranged on the editing timeline.
All of this can be done by importing all of the materials needed for the final edit into a single work session of the software, which can then be selected and sorted into the clipboard for later use in the desired order in the final edit by drag and drop on the timeline.

I have considered alternative solutions with Audacity, such as saving in new files each selected part to reuse it later. But this is a very inefficient and time consuming solution.
I tested of course the use of more complex software and DAW (Ardour and others): none of them has this function and in general using a DAW designed for musicians to edit a journalistic project does not make much sense. This software are way to complex for this use.

What I'm proposing for Audacity is that selected parts of a clip can be copied and pasted directly into a clip panel so that you don't have to go through an export with the creation of a new file, but that only a symbolic reference to a time interval on the clip is actually shown on the panel.
In this way these virtual clips can then be rearranged on the timeline when needed.
This has three advantages in my opinion that are very big:
1.It saves work by avoiding the export step making the workflow more efficient.
2. The result of the export is a file with fixed in and out point, which when put back on the timeline cannot be stretched or shortened according to editing needs (transitions, etc.).With a virtual reference to a part of a clip instead this can be done.
3. Improve the workflow by letting you select pieces of clips and sort them on the clipboard for later use, greatly easing the editing task by giving you more flexibility and choice in the course of your work.

Having a similar feature in Audacity would make this software functional for journalistic work and a nore powerful tool.

Thanks

Mattia

PS: As a reference i propose a screenshot of a software used by my radio, it's from David Systems. Very simple but efficient. On the right the clips panel. And also a screenshot of an audio project of mine in Hindenburg, where you can see the use i do of the clipboard, on the right.

![MultiTrack+screen](https://user-images.githubusercontent.com/97319147/150510998-24115aab-b89d-4fd0-bd39-d43887edb26e.png)
![253724283_10225990233763098_3392269943915397774_n](https://user-images.githubusercontent.com/97319147/150526455-e457a52e-fbb3-48ed-9048-384f1ca6b969.jpg)

**Comment by Paul-Licameli** [comment]
Do I understand that the best workflow you could devise in Audacity is to export each clip to another file?

But I would think that is unnecessary.  You can open another Audacity project with Ctrl + N, copy clips from one project, and paste them into another.  You can keep all the clips in memory at one time.  You can save just one .aup3 file with the clips if you need to save work and resume it later.

You might even avoid the second Audacity project by using a new track to hold the clips, and simply mute it so it is not part of mix-down.

Imperfect, yes, but it would avoid the many trips to the file opening and closing dialogs.

**Comment by Nasosecco** [comment]
Thank you for your answer Paul-Licameli.
I actually tried these two modes you suggest as well. Both of them have a problem (especially the second one). In the first case the problem lies in the fact that there is no place outside the timeline on which to organize the excerpts from the clips, renaming them as needed and perhaps grouping them into sub-folders (e.g. Music, interview one, two, etc.). As you can see from the screenshot of Hindenburg that I posted a project quickly becomes very complicated as well as finding the chunks of a clip.
In the second case that you propose, the project would become even more difficult to manage. With a clipboard to organize and rename the chunks of clips you would have everything in sight and the editing process would be greatly simplified and streamlined.

**Comment by LWinterberg** [comment]
Project bins are a standard feature across most creation software and we should totally build it IMHO. Though my feeling is that it's more of a post-replatforming feature

**Comment by Paul-Licameli** [comment]
> Project bins are a standard feature across most creation software and we should totally build it IMHO. Though my feeling is that it's more of a post-replatforming feature

I sometimes fail to understand the judgments about would should and should not wait for replatforming to Qt.

**Comment by werame** [comment]
> In the programs I normally use (HIndenburg Journalist, DigaSystem) it is possible to select directly in the waveform a part of an interview and add it to a clipboard where all the clips of an interview can be organized to be reused later and rearranged on the editing timeline.
All of this can be done by importing all of the materials needed for the final edit into a single work session of the software, which can then be selected and sorted into the clipboard for later use in the desired order in the final edit by drag and drop on the timeline.

(Adobe) Audition works fairly similarly, as do many DAWs in that regard. There is a catch with that approach the the clips in the multitrack view (to use Adobe's terminology) are merely *references* to parts of waveform. In Auditiona as in many DAWs, you change the waveform, it affects all references to it, unless you use a "make unique" feature to create explicit copies. Because of this referencing model, in Audition it's possible to even slide/reselect the portion included while in the multitrack view with the ["slip tool"](https://www.youtube.com/watch?v=8WRKnhsX1bE).

All that adds complexity that the average clean-my-podcast user might not want. So, Audacity simply doesn't support this (indirect) referencing model. It opted for "what you see it what you get" in terms of waveforms and multi-track not needing two editors in the same program. Which has advantages and disadvantages.

**Comment by werame** [comment]
@Paul-Licameli  If I'm to 2nd guess what LWinterberg is thinking any substantial UI additions (like a way to have per-project external resources window or docklet "bin") would be a waste of effort now if one is committed to abandon wx. It's also not clear to me how easy would be to have that done in wx... apparently there's [wxAUI](https://docs.wxwidgets.org/3.0/overview_aui.html), but I think Audacity isn't using that. If I grep for wxAuiToolBar in Audacity sources, I get nothing.

**Comment by Nasosecco** [comment]
(Adobe) Audition works fairly similarly, as do many DAWs in that regard. There is a catch with that approach that the clips in the multitrack view (to use Adobe's terminology) are merely _references_ to parts of waveforms (from the "library"). In Audition as in many DAWs, if you change the waveform, it affects all references to it, unless you use a "make unique" feature to create explicit copies. Because of this referencing model, in Audition it's possible to even slide/reselect the portion included while in the multitrack view with the ["slip tool"](https://www.youtube.com/watch?v=8WRKnhsX1bE).
> 
> All that adds complexity that the average clean-my-podcast user might not want. So, Audacity simply doesn't support this (indirect) referencing model. It opted for "what you see it what you get" in terms of waveforms and multi-track not needing two editors in the same program. Which has advantages and disadvantages.
> 
> Also, not all DAWs are like Audition. Ableton is bit more similar to Audacity in this regard in the sense that whenever you edit something it automatically makes it unique, essentially doing a copy-on-write. Ardour is somewhere on the fence in that it [gives](https://manual.ardour.org/working-with-midi/copy-midi-region/) the user the choice what the default behavior should be. But still basically has to support he more complex, explicit referencing model. On the other hand, Reaper and FL don't even have an option to make everything unique.

Thank you @werame for your explanation.
I'm not a developer but I understand the reasoning behind Audacity's philosophy, "what you see is what you get". 
I also realize that my proposal identifies potential users who are somewhere between the casual user, who uses Audacity to make a simple cut to a file or apply an effect, and the power user represented by the musician who needs a complete DAW. At the moment there isn't a proposal in the Open Source area (but not even among paid software) that meets this type of user. Apart, as already mentioned, Hindenburg, which is a program much less powerful than Audacity, but whose real selling point lies in its ability to be simpler than a DAW but to provide a tool (the clipboard) to organize the material collected for editing.

Mattia

**Comment by LWinterberg** [comment]
A project bin definitely shines more the better Audacity's support for massively complex projects gets (hence my feeling that this is more a 4.x thing; I mostly see us add features relevant to single-track operations and cleaning up existing janky UI for 3.x). 

>  In Audition as in many DAWs, if you change the waveform, it affects all references to it, unless you use a "make unique" feature to create explicit copies.

I wonder if this is solvable without having to teach the concept of references/clones/instances vs copies to the user.

**Comment by rbdannenberg** [comment]
I also wonder if using key commands or even some custom macros could allow you to save off audio selections to files. File browsers already give tree-structured, scrollable views of collections of files, and you can drag-and-drop files back into Audacity. You could also use an icon view to organize clips spatially in a browser window, and (on Mac) just selecting a file and typing the space bar will preview the audio content. It's probably not as integrated as a clipboard constructed specifically for your workflow, but maybe it's good enough, and files have their own advantages, including possibilities for search, backups, batch processing with Audacity and other tools, etc.

**Comment by werame** [comment]
> I wonder if this is solvable without having to teach the concept of references/clones/instances vs copies to the user.

Well, yes, the Ableton solution is to auto-make everything unique, without asking.

**Comment by Paul-Licameli** [comment]
> @Paul-Licameli If I'm to 2nd guess what LWinterberg is thinking any substantial UI additions (like a way to have per-project external resources window or docklet "bin") would be a waste of effort now if one is committed to abandon wx. It's also not clear to me how easy would be to have that done in wx... apparently there's [wxAUI](https://docs.wxwidgets.org/3.0/overview_aui.html), but I think Audacity isn't using that. If I grep for wxAuiToolBar in Audacity sources, I get nothing. Being able to dock the macros window in present Audacity would be quite helpful, by the way. (I guess I can make an enhancement request for that 🧀 ) It's a standard feature in competitors like Audition; which actually comes with it opened and pre-dock, albeit it's called "favorites"... but a "favorite" can store multi-step sequences or user actions, so they're really macros.
> 
> I know that Dominic Mazzoni had this vision of a really simple program and that's probably one of the reasons why Audacity took off, so I guess that he would disapprove of the Audition approach to show you the macros/favorites window from the start, and that he might also disapprove of showing similar complexity like a (project) resources library, by default. The fact that the macros window in Audacity is non-modal is good, but that is also not a totally obvious feature.

I strongly disagree with the opinion that all of this would be wasted effort if it were attempted before replatforming.  Honestly, that is an ill-informed hunch and a non-sequitur if developers aren't saying it.  I say don't listen.

Are you offering to explore the problem @werame ?  I say, go for it.  There are much more intrinsic difficulties to figure out, than these minor details of UI, and whether the UI is in wxWidgets or Qt is neither here nor there.  That more major part of the effort, if successful, will not be wasted.

Try it first as a non-modal dialog.  Follow the established pattern of the other non-modal dialogs (Undo history, Contrast, Plot Spectrum, Lyrics, Mixer Board, Macros).  I think something acceptable and useful, just with dialog controls and without fancy graphics, could be devised.

Having solved all the other problems that must solved by our team, to conserve all of the other application logic with a migrated user interface -- then I think this feature would not add any new problems.  Only some more instances of the common problems which will become routinized by then.

Using a notebook control like wxAuiNotebook, as an alternative to a non-modal dialog, might also prove easier than expected.  I think it isn't yet done in Audacity merely because it just wasn't tried.  That too might be worth the trial, even just for the existing non-modal dialogs, though there is less of a component of platform-independent core data structure work there.

My hope for replatforming -- I don't speak with authority -- is that it will be incremental, experimental, laboratory work that proceeds beside the regular 3.x releases of the working wxWidgets program, which will have improvements for the user.  This concurrent development makes more sense to me, given the large uncertainties around the big undertaking and the wide cone of uncertainty now about when it could complete.  I hope it will not be a stop-the-world hiatus in all other development.

**Comment by werame** [comment]
Sorry, no, I can't help code this. Anything with extensive UI would have to go through Muse corp UX design meetings etc.

**Comment by LWinterberg** [comment]
> If I'm to 2nd guess what LWinterberg is thinking any substantial UI additions (like a way to have per-project external resources window or docklet "bin") would be a waste of effort now if one is committed to abandon wx.

No, it's got nothing to do with it being a waste or UI additions being too difficult. It's just about how I see the larger-scale roadmap. Audacity needs many improvements in many areas, with this being one of the ones we definitely need to add. But what priority does it have relative to other improvements? 

I think a project bin is not as important as fixing up existing tools that are underutilized due to bad discoverability or design, or adding features that massively improve single-track editing. I also don't think it's as important as replatforming itself.

In other words, I think it's important to streamline existing use cases and workflows first before introducing new workflows. Project bins are key to enabling new workflows, but I don't see them massively improving existing ones - if you *need* a project bin for your work or else need to rely on various amounts of ugly workarounds, you *probably* are using another tool if you actually need to get work done - as is evidenced by OP.

**Comment by Paul-Licameli** [comment]
> > If I'm to 2nd guess what LWinterberg is thinking any substantial UI additions (like a way to have per-project external resources window or docklet "bin") would be a waste of effort now if one is committed to abandon wx.
> 
> No, it's got nothing to do with it being a waste or UI additions being too difficult. It's just about how I see the larger-scale roadmap. Audacity needs many improvements in many areas, with this being one of the ones we definitely need to add. But what priority does it have relative to other improvements?
> 
> I think a project bin is not as important as fixing up existing tools that are underutilized due to bad discoverability or design, or adding features that massively improve single-track editing. I also don't think it's as important as replatforming itself.
> 
> In other words, I think it's important to streamline existing use cases and workflows first before introducing new workflows. Project bins are key to enabling new workflows, but I don't see them massively improving existing ones - if you _need_ a project bin for your work or else need to rely on various amounts of ugly workarounds, you _probably_ are using another tool if you actually need to get work done - as is evidenced by OP.

That is a fair summary of priorities we on the team will set, to do the highest valued things for the user community.  But I'm also saying, if a contributor has time time and ability and motivation to find a good solution to this problem, then that's welcome too and not wasted effort even if it's not yet based on Qt.

**Comment by SteveDaulton** [comment]
> I think a project bin is not as important as fixing up existing tools that are underutilized due to bad discoverability or design, or adding features that massively improve single-track editing.

or making Audacity less buggy so that it can be used with confidence for real work.

**Comment by werame** [comment]
I have to make a correction to what I said earlier. With Audacity 3.1 "smart clips", it's not always what you see is what you get... #2435

**Comment by LWinterberg** [comment]
this is planned for much later (Audacity 4.x), closing for now

**Comment by SteveDaulton** [comment]
> * Selection in each interview of the parts that will be used in the final edit
> * Dividing the interview into segments that will then be organized along with music and sound effects in the final edit

The way that I usually handle that kind of task is to put my "source" tracks at the top of the project and mute them. Then assemble the "production" tracks below.

Some people prefer to have their "source" tracks in a separate project window, and copy / paste audio across to the "production" project as required, but I generally prefer to work all in one project.


I think it would also be possible to implement an "Advanced Clipboard" as a Nyquist plug-in. If you are interested in working with me on such a plug-in, please create a new topic in this section of the Audacity forum: https://forum.audacityteam.org/viewforum.php?f=42 (that's where I usually hang out).

**Comment by Nasosecco** [comment]
Hello @SteveDaulton, 
Thank you for your proposal, but unfortunately I am not a programmer, so I am not sure how I could help....
Let me know if I can do anything.
